src/lib/ServoOutput/waveform_8266.cpp

```diff
@@ -0,0 +1,322 @@
+ *   cycle without blocking.
+ * - Removes analogWrite() functionality. If this is used, it will break
+ *   this code, as the standard core_esp8266_waveform_pwm will take over
+ *   the timer.
+ ***/
+/*
+  esp8266_waveform - General purpose waveform generation and control,
+                     supporting outputs on all pins in parallel.
+
+  Copyright (c) 2018 Earle F. Philhower, III.  All rights reserved.
+
+  The core idea is to have a programmable waveform generator with a unique
+  high and low period (defined in microseconds or CPU clock cycles).  TIMER1
+  is set to 1-shot mode and is always loaded with the time until the next
+  edge of any live waveforms.
+
+  Up to one waveform generator per pin supported.
+
+  Each waveform generator is synchronized to the ESP clock cycle counter, not
+  the timer.  This allows for removing interrupt jitter and delay as the
+  counter always increments once per 80MHz clock.  Changes to a waveform are
+  contiguous and only take effect on the next waveform transition,
+  allowing for smooth transitions.
+
+  This replaces older tone(), analogWrite(), and the Servo classes.
+
+  Everywhere in the code where "cycles" is used, it means ESP.getCycleCount()
+  clock cycle count, or an interval measured in CPU clock cycles, but not
+  TIMER1 cycles (which may be 2 CPU clock cycles @ 160MHz).
+
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+*/
+
+
+#include <Arduino.h>
+#include "ets_sys.h"
+
+// Maximum delay between IRQs
+#define MAXIRQUS (10000)
+
+// Waveform generator can create tones, PWM, and servos
+typedef struct {
+  uint32_t nextServiceCycle;   // ESP cycle timer when a transition required
+  uint32_t timeHighCycles;     // Actual running waveform period (adjusted using desiredCycles)
+  uint32_t timeLowCycles;      //
+  uint32_t desiredHighCycles;  // Ideal waveform period to drive the error signal
+  uint32_t desiredLowCycles;   //
+  uint32_t nextHighLowUs;      // Waveform ideal (us) "on deck", waiting to be changed next cycle
+                               // packed into 32 bits to be atomic read/write, 65535us max
+  uint32_t lastEdge;           // Cycle when this generator last changed
+} Waveform;
+
+class WVFState {
+public:
+  Waveform waveform[17];        // State of all possible pins
```

Comment:
Is 17 related to a hw limit or to the number of channels?

Reply:
GPIO0-16 on the esp8266 so I think theres a slot for each possible pin.

Reply:
That's correct. The code came from the esp8286 core except it only can change one servo per edge and has no way of accessing its internals to change that. I had to copy paste it over but then I cut out a lot of the other complexity. The RAM usage is about 10 more bytes overall due to each channel having to store its own next-queued-value. It really should be the max pin number we plan to use, but I figured that might be a hidden gotcha somewhere down the line if someone tried to use a pin higher than we use. Adding double indirection (a list of pins to waveform mapping) added a lot of code everywhere so I just left it how it was in the core at 17.